File: cpp/yolo.cpp
/* 
 * @Author: taifyang
 * @Date: 2025-10-21 06:50:02
 * @LastEditTime: 2026-08-29 22:54:23
 * @Description: source file for YOLO algorithm
 */

#include <chrono>
#include "yolo.h"
#include "utils.h"


#ifdef _YOLO_LIBTORCH
	#include "libtorch/yolo_libtorch.h"
#endif // _YOLO_Libtorch

#ifdef _YOLO_ONNXRUNTIME
	#include "onnxruntime/yolo_onnxruntime.h"
#endif // _YOLO_ONNXRuntime

#ifdef _YOLO_OPENCV
	#include "opencv/yolo_opencv.h"
#endif // _YOLO_OpenCV

#ifdef _YOLO_OPENVINO
	#include "openvino/yolo_openvino.h"
#endif // _YOLO_OpenVINO

#ifdef _YOLO_TENSORRT
	#include "tensorrt/yolo_tensorrt.h"
#endif // _YOLO_TensorRT


void YOLO::infer(const std::string file_path, bool save_result, bool show_result, char* argv[])
{
	if(!std::filesystem::exists(file_path))
	{
		std::cerr << "file not exists!" << std::endl;
		std::exit(-1);
	}

	m_draw_result = save_result || show_result;
	std::string extension = std::filesystem::path(file_path).extension().string();
	if (extension == ".bmp" || extension == ".jpg" || extension == ".png")
	{
		m_image = cv::imread(file_path);
		if (m_image.empty())
		{
			std::cerr << "read image empty!" << std::endl;
			std::exit(-1);
		}

		//warm up
		for(int i=0; i<10; ++i)
		{
			pre_process();
			process();
			post_process();
		}	

		auto start = std::chrono::steady_clock::now();
		for(int i=0; i<1000; ++i)
		 {
		  	pre_process();
		  	process();
		  	post_process();
		 }		
		 auto end = std::chrono::steady_clock::now();	
		 std::chrono::duration<double> duration = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);		
		std::cout << "avg cost run on 1000 times:" << duration.count() << "ms" << std::endl;

		if (save_result)
		{
			std::string result_name = std::string(argv[1]) + "_" + std::string(argv[2]) + "_" +std::string(argv[3]) + "_"  + std::string(argv[4]) + "_"  + std::string(argv[5]) + ".jpg";
			cv::imwrite(result_name, m_result);
		}
		if (show_result)
		{
			cv::imshow("result", m_result);
			cv::waitKey(0);
			cv::destroyAllWindows();
		}
	}
	else if (extension == ".mp4")
	{
		cv::VideoCapture cap;
		if (!cap.open(file_path))
		{
			std::cerr << "cannot read video!" << std::endl;
			std::exit(-1);
		}

		cv::VideoWriter wri;
		if (save_result)
		{
			std::string result_name = "./result/" + std::string(argv[1]) + "_" + std::string(argv[2]) + "_" +std::string(argv[3]) + "_"  + std::string(argv[4]) + "_"  + std::string(argv[5]) + ".avi";
			int width = int(cap.get(cv::CAP_PROP_FRAME_WIDTH));
    		int height = int(cap.get(cv::CAP_PROP_FRAME_HEIGHT));
			int fps = int(cap.get(cv::CAP_PROP_FPS));
			wri.open(result_name, cv::VideoWriter::fourcc('M', 'P', '4', 'V'), fps, cv::Size(width, height));
		}

		auto start = std::chrono::steady_clock::now();

		while (true)
		{
			cap.read(m_image);
			if (m_image.empty())
			{
				break;
			}

			pre_process();
			process();
			post_process();

			if (save_result)
			{
				wri << m_result;
			}
			if (show_result)
			{
				cv::imshow("result", m_result);
				cv::waitKey(1);
			}
		}	

		auto end = std::chrono::steady_clock::now();	
		std::chrono::duration<double> duration = std::chrono::duration_cast<std::chrono::duration<double>>(end - start);		
		std::cout << duration.count() * 1000 << "ms" << std::endl;
		cap.release();

		if (save_result)
		{
			wri.release();
		}
		if (show_result)
		{
			cv::destroyAllWindows();
		}
	}
	else
	{
		std::cerr << "supported input file types: .bmp .jpg .png .mp4" << std::endl;
		std::exit(-1);
	}
}

CreateFactory& CreateFactory::instance()
{
	static CreateFactory create_factory;
	return create_factory;
}

void CreateFactory::register_class(const Backend_Type& backend_type, const Task_Type& task_type, CreateFunction create_function)
{
	m_create_registry[backend_type][task_type] = create_function;
}

std::unique_ptr<YOLO> CreateFactory::create(const Backend_Type& backend_type, const Task_Type& task_type)
{
	std::unique_ptr<YOLO> yolo = m_create_registry[backend_type][task_type]();
	if(yolo == nullptr)
	{
		std::cerr << "algo create failed!" <<std::endl;
		std::exit(-1);
	}
	else 
	{
		return yolo;
	}
}

CreateFactory::CreateFactory()
{
    size_t backend_size = magic_enum::enum_count<Backend_Type>();
    size_t task_size = magic_enum::enum_count<Task_Type>();
    m_create_registry.resize(backend_size, std::vector<CreateFunction>(task_size));

#ifdef _YOLO_LIBTORCH
#  define X(TkStr,TkEnum) register_class(Backend_Type::Libtorch, TkEnum, +[]()->std::unique_ptr<YOLO>{return std::make_unique<YOLO_Libtorch_##TkStr>();});
#else
#  define X(TkStr,TkEnum) register_class(Backend_Type::Libtorch, TkEnum, +[]()->std::unique_ptr<YOLO>{return nullptr;});
#endif
YOLO_TASK_LIST
#undef X

#ifdef _YOLO_ONNXRUNTIME
#  define X(TkStr,TkEnum) register_class(Backend_Type::ONNXRuntime, TkEnum, +[]()->std::unique_ptr<YOLO>{return std::make_unique<YOLO_ONNXRuntime_##TkStr>();});
#else
#  define X(TkStr,TkEnum) register_class(Backend_Type::ONNXRuntime, TkEnum, +[]()->std::unique_ptr<YOLO>{return nullptr;});
#endif
YOLO_TASK_LIST
#undef X

#ifdef _YOLO_OPENCV
#  define X(TkStr,TkEnum) register_class(Backend_Type::OpenCV, TkEnum, +[]()->std::unique_ptr<YOLO>{return std::make_unique<YOLO_OpenCV_##TkStr>();});
#else
#  define X(TkStr,TkEnum) register_class(Backend_Type::OpenCV, TkEnum, +[]()->std::unique_ptr<YOLO>{return nullptr;});
#endif
YOLO_TASK_LIST
#undef X

#ifdef _YOLO_OPENVINO
#  define X(TkStr,TkEnum) register_class(Backend_Type::OpenVINO, TkEnum, +[]()->std::unique_ptr<YOLO>{return std::make_unique<YOLO_OpenVINO_##TkStr>();});
#else
#  define X(TkStr,TkEnum) register_class(Backend_Type::OpenVINO, TkEnum, +[]()->std::unique_ptr<YOLO>{return nullptr;});
#endif
YOLO_TASK_LIST
#undef X

#ifdef _YOLO_TENSORRT
#  define X(TkStr,TkEnum) register_class(Backend_Type::TensorRT, TkEnum, +[]()->std::unique_ptr<YOLO>{return std::make_unique<YOLO_TensorRT_##TkStr>();});
#else
#  define X(TkStr,TkEnum) register_class(Backend_Type::TensorRT, TkEnum, +[]()->std::unique_ptr<YOLO>{return nullptr;});
#endif
YOLO_TASK_LIST
#undef X

#undef YOLO_TASK_LIST
}

File: python/backends/yolo.py
'''
Author: taifyang 
Date: 2024-06-12 22:23:07
LastEditors: taifyang 58515915+taifyang@users.noreply.github.com
LastEditTime: 2024-11-30 11:20:18
FilePath: \python\backends\yolo.py
Description: YOLO algorithm interface class
'''


import os
import cv2
import time
import backends
    

'''
description: YOLO algorithm interface class
'''
class YOLO:  
    '''
    description:    construction method
    param {*} self  instance of class
    return {*}
    '''    
    def __init__(self) -> None:
        super().__init__()
        self.class_num = 80		            
        self.score_threshold = 0.2      	
        self.nms_threshold = 0.5        	
        self.confidence_threshold = 0.2 	
        self.inputs_shape = (640, 640)   	

    '''
    description:    task map
    param {*} self  instance of class
    return {*}      algorithm class instance
    '''    
    def task_map(self):
        map = {}
        try:
            map['ONNXRuntime'] = {
                'Classify':backends.ONNXRuntime.YOLO_ONNXRuntime_Classify,
                'Detect':backends.ONNXRuntime.YOLO_ONNXRuntime_Detect,
                'Segment':backends.ONNXRuntime.YOLO_ONNXRuntime_Segment,
            }
        except:
               pass
        
        try:
            map['OpenCV'] =  {
                'Classify':backends.OpenCV.YOLO_OpenCV_Classify,
                'Detect':backends.OpenCV.YOLO_OpenCV_Detect,
                'Segment':backends.OpenCV.YOLO_OpenCV_Segment,
            }
        except:
            pass
                             
        try:
            map['OpenVINO'] = {
                'Classify':backends.OpenVINO.YOLO_OpenVINO_Classify,
                'Detect':backends.OpenVINO.YOLO_OpenVINO_Detect,
                'Segment':backends.OpenVINO.YOLO_OpenVINO_Segment,
            }
        except:
            pass
        
        try:
            map['PyTorch'] =  {
                'Classify':backends.PyTorch.YOLO_PyTorch_Classify,
                'Detect':backends.PyTorch.YOLO_PyTorch_Detect,
                'Segment':backends.PyTorch.YOLO_PyTorch_Segment,
            }
        except:
            pass
        
        try:
            map['TensorRT'] = {
                'Classify':backends.TensorRT.YOLO_TensorRT_Classify,
                'Detect':backends.TensorRT.YOLO_TensorRT_Detect,
                'Segment':backends.TensorRT.YOLO_TensorRT_Segment,
            }
        except:
            pass

        return map
    
    '''
    description:                inference interface
    param {*} self              instance of class
    param {str} input_path      input path
    param {str} output_path     output path
    param {bool} save_result    save result
    param {bool} show_result    show result
    return {*}
    '''    
    def infer(self, input_path:str, output_path:str, save_result:bool, show_result:bool) -> None:
        assert os.path.exists(input_path), 'input not exists!'
        self.draw_result = save_result or show_result
        if input_path.endswith('.bmp') or input_path.endswith('.jpg') or input_path.endswith('.png'):
            self.image = cv2.imread(input_path)
            self.result = self.image.copy()
        
            #warm up
            for i in range(10):
                self.pre_process()
                self.process()
                self.post_process()
            
            start = time.perf_counter()
            for i in range(1000):
                self.pre_process()
                self.process()
                self.post_process()
            end = time.perf_counter()
            print('avg cost run on 1000 times:', end-start, 'ms')  
            
            if save_result and output_path!='':
                cv2.imwrite(output_path, self.result)
            if show_result:
                cv2.imshow('result', self.result)
                cv2.waitKey(0)
        elif input_path.endswith('.mp4'):
            cap = cv2.VideoCapture(input_path)
            start = time.perf_counter()
            if save_result and output_path!='':
                fourcc = cv2.VideoWriter_fourcc(*'XVID')
                fps = cap.get(cv2.CAP_PROP_FPS)
                width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                wri = cv2.VideoWriter(output_path, fourcc, fps, (width,height))
            while True:
                ret, self.image  = cap.read()
                if not ret:
                    break
                self.result = self.image.copy()
                self.pre_process()
                self.process()
                self.post_process()
                if show_result:
                    cv2.imshow('result', self.result)
                    cv2.waitKey(1)
                if save_result and output_path!='':
                    wri.write(self.result)
            end = time.perf_counter()
            print((end-start)*1000, 'ms')                         
            